### PR TITLE
refactor ChunkedPolynomials

### PR DESCRIPTION
**Pull request by xiangxiecrypto**
This PR (related to issue #432):
- add a new module  in `utils/chunked_polynomial.rs` with an abstract struct `ChunkedPolynomials`.
- the new `ChunkedPolynomials` struct contains two method: `linearize` and `compress_polynomial`. `linearize` evaluates all the chunked polynomials in the same point. `compress_polynomial` compresses all the chunked polynomials into one with `zeta_n`.
- refactor the original struct `ChunkedPolynomial` with `ScaledChunkedPolynomials` based on the above abstract. This struct contains `(Vec[F], Vec[P])`.
-  add a function `to_chunked_polynomials` in trait `ExtendedDensePolynomial`, which allows to transfer polynomials into multiple chunks.
- modify accordingly in the `prover.rs` file.

### CODE REVIEW

File: utils/src/chunked_polynomial.rs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+    pub fn add_shifted(&mut self, scale: F, shift: usize, p: P) {
+        self.scale.push(scale);
+        self.chunked_polynomials
+            .add_chunk(OptShiftedPolynomial::Shifted(p, shift))
+    }
+}
+
+impl<'a, F: Field> ScaledChunkedPolynomials<F, &'a [F]> {
+    /// check length?
```
Comment (querolita) [inline utils/src/chunked_polynomial.rs:101]: if this is a TODO, change to normal comment // 
and if it is indeed a TODO, is there any test to check the length? it seems like an important check to make sure dense polynomials will be correctly formed

File: utils/src/dense_polynomial.rs [hunk 2]
```diff
@@ -19,14 +21,7 @@ pub trait ExtendedDensePolynomial<F: Field> {
...
 
-    /// This function evaluates polynomial in chunks.
-    fn eval(&self, elm: F, size: usize) -> Vec<F>;
-
-    /// Multiplies the chunks of a polynomial with powers of zeta^n to make it of degree n-1.
-    /// For example, if a polynomial can be written `f = f0 + x^n f1 + x^2n f2`
-    /// (where f0, f1, f2 are of degree n-1), then this function returns the new semi-evaluated
-    /// `f'(x) = f0(x) + zeta^n f1(x) + zeta^2n f2(x)`.
-    fn chunk_polynomial(&self, zeta_n: F, n: usize) -> Self;
+    fn to_chunked_polynomials(&self, size: usize) -> ChunkedPolynomials<DensePolynomial<F>>;
```
[the remark this answers is not in the document]
Comment (mimoo) [reply utils/src/dense_polynomial.rs:24]: `to_chunked_polynomial` singular no?

File: utils/src/chunked_polynomial.rs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        }
+
+        while coeffs.last().map_or(false, |c| c.is_zero()) {
+            coeffs.pop();
+        }
+
+        DensePolynomial { coeffs }
+    }
+}
+enum OptShiftedPolynomial<P> {
```
Comment (mimoo) [inline utils/src/chunked_polynomial.rs:61]: nit: missing space between the two structs

File: utils/src/chunked_polynomial.rs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    Unshifted(P),
+    Shifted(P, usize),
+}
+
+/// A formal sum of the form
+/// `s_0 * p_0 + ... s_n * p_n`
+/// where each `s_i` is a scalar and each `p_i` is an optionally shifted polynomial.
```
Comment (mimoo) [inline utils/src/chunked_polynomial.rs:68]: this comment seems unrelated to the struct below

File: utils/src/chunked_polynomial.rs (hunk 3 above)
Comment (mimoo) [reply utils/src/chunked_polynomial.rs:61]: also some doc would be useful

File: utils/src/chunked_polynomial.rs [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+///pub struct ChunkedPolynomial<F, P>(Vec<(F, OptShiftedPolynomial<P>)>);
+
+pub struct ScaledChunkedPolynomials<F, P> {
+    scale: Vec<F>,
+    chunked_polynomials: ChunkedPolynomials<OptShiftedPolynomial<P>>,
+}
+
+impl<F, P> Default for ScaledChunkedPolynomials<F, P> {
+    fn default() -> ScaledChunkedPolynomials<F, P> {
```
Comment (mimoo) [inline utils/src/chunked_polynomial.rs:78]: derive Default on the struct instead

File: utils/src/chunked_polynomial.rs [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+use rayon::prelude::*;
+
+#[derive(Clone)]
+pub struct ChunkedPolynomials<P> {
+    pub polys: Vec<P>,
+    pub degree: usize,
+}
+
+impl<P> Default for ChunkedPolynomials<P> {
```
[the remark this answers is not in the document]
Comment (mimoo) [reply utils/src/chunked_polynomial.rs:12]: also, should this really have a default implementation? It sounds to me like you would always create this struct from a `DensePolynomial`

File: utils/src/chunked_polynomial.rs [hunk 7]
```diff
@@ -0,0 +1,164 @@
+use ark_ff::{Field, Zero};
+use ark_poly::polynomial::{univariate::DensePolynomial, Polynomial, UVPolynomial};
+
+use rayon::prelude::*;
+
+#[derive(Clone)]
+pub struct ChunkedPolynomials<P> {
+    pub polys: Vec<P>,
+    pub degree: usize,
```
Comment (mimoo) [inline utils/src/chunked_polynomial.rs:9]: both the struct and the fields need documentation. Here `degree` sounds like it could be renamed to `chunk_degree` or something

File: utils/src/chunked_polynomial.rs [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+/// A formal sum of the form
+/// `s_0 * p_0 + ... s_n * p_n`
+/// where each `s_i` is a scalar and each `p_i` is an optionally shifted polynomial.
+
+///pub struct ChunkedPolynomial<F, P>(Vec<(F, OptShiftedPolynomial<P>)>);
+
+pub struct ScaledChunkedPolynomials<F, P> {
+    scale: Vec<F>,
+    chunked_polynomials: ChunkedPolynomials<OptShiftedPolynomial<P>>,
+}
```
Comment (mimoo) [inline utils/src/chunked_polynomial.rs:75]: I'm a bit confused what this struct is about. (I know you didn't add it.) Same comment as previously, I think this should be `Polynomial` singular. What puzzles me is why we have several shifted polynomials, when we usually only care about the last one (the shift being interesting for upperbounding the degree of a poly)

File: utils/src/lib.rs [hunk 9]
```diff
@@ -1,6 +1,8 @@
+pub mod chunked_polynomial;
 pub mod dense_polynomial;
 pub mod evaluations;
 pub mod field_helpers;
+
```
Comment (mimoo) [inline utils/src/lib.rs:5]: remove the space

**Review by mimoo — CHANGES_REQUESTED** [review]
I'm mid-review, making sure that this doesn't get merged as I think it still needs some work.

That being said, I really like the direction so far :) I think this is going to greatly improve the codebase.

File: utils/src/chunked_polynomial.rs [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+    }
+}
+
+impl<P> ChunkedPolynomials<P> {
+    pub fn add_chunk(&mut self, p: P) {
+        self.polys.push(p)
+    }
+}
+
+impl<F: Field> ChunkedPolynomials<DensePolynomial<F>> {
```
Comment (mimoo) [inline utils/src/chunked_polynomial.rs:27]: I'm not a fan of this generic parameter. I think it's make more sense to define `ChunkedPolynomial` as

```rust
pub struct ChunkedPolynomial<F> {
    pub polys: Vec<DensePolynomial<F>>,
    pub degree: usize,
}
```

as this is what we're going to use it with mostly.

File: poly-commitment/src/commitment.rs [hunk 11]
```diff
@@ -424,65 +423,6 @@ pub fn combined_inner_product<G: CommitmentCurve>(
...
-            .collect();
-
-        for p in scaled {
-            res += &p;
-        }
-
-        res
-    }
-}
-
```
Comment (mimoo) [inline poly-commitment/src/commitment.rs:485]: I actually think we should try not to touch this part of the code (poly-commitment) yet. As this logic seems a bit unnecessary to me at the moment. I think this PR could be more useful if targeted to the kimchi crate to begin with.

File: utils/src/dense_polynomial.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (querolita) [reply utils/src/dense_polynomial.rs:24]: Oh way better indeed

File: utils/src/chunked_polynomial.rs [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    /// This function evaluates polynomial in chunks.
+    pub fn linearize(&self, elm: F) -> Vec<F> {
+        let mut res: Vec<F> = vec![];
+        for poly in self.polys.iter() {
+            let eval = poly.evaluate(&elm);
+            res.push(eval);
+        }
+        res
+    }
```
[the remark this answers is not in the document]
Comment (querolita) [reply utils/src/chunked_polynomial.rs:36]: good point

File: utils/src/chunked_polynomial.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (xiangxiecrypto) [reply utils/src/chunked_polynomial.rs:12]: > also, should this really have a default implementation? It sounds to me like you would always create this struct from a `DensePolynomial`

This is for `ScaledChunkedPolynomial` , the original `ChunkedPolynomial` used in opening proof.

File: utils/src/chunked_polynomial.rs [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+            for (coeff, poly_coeff) in coeffs.iter_mut().zip(&poly.coeffs) {
+                *coeff += scale * poly_coeff;
+            }
+
+            scale *= zeta_n;
+        }
+
+        while coeffs.last().map_or(false, |c| c.is_zero()) {
+            coeffs.pop();
+        }
```
[the remark this answers is not in the document]
Comment (xiangxiecrypto) [reply utils/src/chunked_polynomial.rs:56]: > I feel like this could be rewritten in a clearer way. Perhaps something like this:
> 
> ```rust
> let coeffs = coeffs.rev().skip_while(|c| c.is_zero()).rev().collect();
> ```

well, I copy the codes here... Yes, the new one is better i think.

File: utils/src/chunked_polynomial.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (xiangxiecrypto) [reply utils/src/chunked_polynomial.rs:36]: > I wouldn't call this function `linearize`. Linearizing a poly to me is:
> 
> given a poly `f(x) = f0(x) + x^n f1(x) + x^2n f2(x) + ...` and a point `z`, return f(x) = f0(x) + z^n f1(x) + z^2n f2(x) + ...`
> 
> where all the `f_i` are of degree `n-1`
> 
> this function is more like `evaluate_chunks` to me

Ok, this means I have to rename `linearize` and `compress_polynomials`

File: utils/src/chunked_polynomial.rs (hunk 10 above)
Comment (xiangxiecrypto) [reply utils/src/chunked_polynomial.rs:27]: > I'm not a fan of this generic parameter. I think it's make more sense to define `ChunkedPolynomial` as
> 
> ```rust
> pub struct ChunkedPolynomial<F> {
>     pub polys: Vec<DensePolynomial<F>>,
>     pub degree: usize,
> }
> ```
> 
> as this is what we're going to use it with mostly.

I thought about this before, the reason I use more generic parameters is to be compatible with the definition of `ScaledChunkedPolynomials`. Perhaps, we could refactor it.

File: poly-commitment/src/commitment.rs (hunk 11 above)
Comment (xiangxiecrypto) [reply poly-commitment/src/commitment.rs:485]: > I actually think we should try not to touch this part of the code (poly-commitment) yet. As this logic seems a bit unnecessary to me at the moment. I think this PR could be more useful if targeted to the kimchi crate to begin with. Would you mind moving the poly-commitment stuff in this PR to another PR/branch for now?

Well, if we refactor `ChunkedPolynomial` into a new module, it is necessary that we move this part into the new file.

File: utils/src/chunked_polynomial.rs [hunk 14]
```diff
@@ -0,0 +1,164 @@
+use ark_ff::{Field, Zero};
+use ark_poly::polynomial::{univariate::DensePolynomial, Polynomial, UVPolynomial};
+
+use rayon::prelude::*;
+
+#[derive(Clone)]
+pub struct ChunkedPolynomials<P> {
```
[the remark this answers is not in the document]
Comment (xiangxiecrypto) [reply utils/src/chunked_polynomial.rs:7]: > wouldn't `ChunkedPolynomial` singular be better here? Since it's a polynomial that's chunked

Yes, that is better.

File: utils/src/chunked_polynomial.rs [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+
+            scale *= zeta_n;
+        }
+
+        while coeffs.last().map_or(false, |c| c.is_zero()) {
+            coeffs.pop();
+        }
+
+        DensePolynomial { coeffs }
+    }
```
[the remark this answers is not in the document]
Comment (xiangxiecrypto) [reply utils/src/chunked_polynomial.rs:59]: > this is the `linearize` function to me :o could be used here for example: https://github.com/o1-labs/marlin/blob/37b7ad53f968c1368ce61ca791aedafe3c06fc33/kimchi/src/prover.rs#L839

Hmmm, maybe I misunderstand the meaning in Issue #432 :), will change that.

File: utils/src/chunked_polynomial.rs (hunk 8 above)
Comment (xiangxiecrypto) [reply utils/src/chunked_polynomial.rs:75]: > I'm a bit confused what this struct is about. (I know you didn't add it.) Same comment as previously, I think this should be `Polynomial` singular. What puzzles me is why we have several shifted polynomials, when we usually only care about the last one (the shift being interesting for upperbounding the degree of a poly)

This struct is used in commitment, where one has to bound the degree in the original implementation.

File: utils/src/chunked_polynomial.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (xiangxiecrypto) [reply utils/src/chunked_polynomial.rs:56]: > I feel like this could be rewritten in a clearer way. Perhaps something like this:
> 
> ```rust
> let coeffs = coeffs.rev().skip_while(|c| c.is_zero()).rev().collect();
> ```

This is not the functionality we want. Here we remove the leading zeros, while the code above remove all the zero coefficient.
Actually, I just copy the codes in the original implementation. :)

**Comment by xiangxiecrypto** [comment]
Update:
- Use less generic definition of `ChunkedPolynomial` with field `Vec<DensePolynomial<F>>`.
- Add documentations and use suitable names for pub structs and functions.
- Move `ScaledChunkedPolynomial` into `evaluation_proof.rs`, the struct and functions are private.

File: utils/src/chunked_polynomial.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (mimoo) [reply utils/src/chunked_polynomial.rs:42]: I think they're all equivalent :o https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e7dda821e6964961bdb6d350929edaac

> Actually, I just copy the codes in the original implementation. :)

yeah sorry about that ^^

File: utils/src/chunked_polynomial.rs [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+        let res = (one + zeta)
+            * (one + zeta_n + zeta_n * zeta.square() + zeta_n * zeta.square() * zeta.square());
+
+        // 1 + x + x^2 + x^3 + x^4 + x^5 + x^6 + x^7 = (1+x) + x^2 (1+x) + x^4 (1+x) + x^6 (1+x)
+        let coeffs = [one, one, one, one, one, one, one, one];
+        let f = DensePolynomial::from_coefficients_slice(&coeffs);
+
+        let eval = f.to_chunked_polynomial(2).linearize(zeta_n).evaluate(&zeta);
+
+        assert!(eval == res);
```
Comment (mimoo) [inline utils/src/chunked_polynomial.rs:72]: this should be equal to `f.evaluate(&zeta)` as well

File: poly-commitment/src/evaluation_proof.rs [hunk 17]
```diff
@@ -12,6 +12,60 @@ use rand_core::{CryptoRng, RngCore};
...
+            .collect();
+
+        for p in scaled {
+            res += &p;
+        }
+
+        res
+    }
+}
+
```
Comment (mimoo) [inline poly-commitment/src/evaluation_proof.rs:68]: I see, ScaledChunkedCommitment is only used here. I think I have some idea on how to get rid of it in the future

**Comment by xiangxiecrypto** [comment]
> 

Let me think about it.

Thanks for the reviews @mimoo @querolita :)
